### PR TITLE
Feat/add azure blob storage support

### PR DESCRIPTION
**Pull request by SimonFrauenschuh**
### Add Support for Azure Blob Storage

Following the code and tests of the Micronaut implementation of the aws S3-client, I've tried to implement something similar for Azure.

**Problem**
- The crud test in class `BlobCustomerControllerTest` fails and I was not able to solve the issue
   - If debugged, it fails because of a timeout of the HTTP-Controller that makes the request
   - If run, it fails after successfully looking up the Azure Storage Account Client with the following error messages 
```
Message: block()/blockFirst()/blockLast() are blocking, which is not supported in thread default-nioEventLoopGroup-2-3
Path Taken: new $CustomerController$Definition$Intercepted(CustomerRepository repository,BeanResolutionContext $beanResolutionContext,BeanContext $beanContext,Qualifier $qualifier,List $interceptors,InterceptorRegistry $interceptorRegistry) --> new $CustomerController$Definition$Intercepted([CustomerRepository repository],BeanResolutionContext $beanResolutionContext,BeanContext $beanContext,Qualifier $qualifier,List $interceptors,InterceptorRegistry $interceptorRegistry) --> new $CustomerRepositoryStoreImpl$Definition$Intercepted([RootProvider rootProvider],BeanResolutionContext $beanResolutionContext,BeanContext $beanContext,Qualifier $qualifier,List $interceptors,InterceptorRegistry $interceptorRegistry) --> new DefaultRootProvider([StorageManager storageManager])
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [io.micronaut.eclipsestore.DefaultRootProvider]
```

```
[default-nioEventLoopGroup-2-3] TRACE i.m.h.s.netty.body.NettyByteBody - Body claimed at this location. This is not an error, but may aid in debugging other errors
java.lang.Exception: null
```

The client needs to be [defined explicitly](https://github.com/micronaut-projects/micronaut-eclipsestore/compare/1.5.x...SimonFrauenschuh:micronaut-eclipsestore:feat/add-azure-blob-storage-support?expand=1#diff-e80965ef0f1a396beb49a195ceb015fe51878fa186b193a61606daf588434baeR38) here because creating the Blob-container in the Azurite-Docker-container is a blocking call and if moved to `AzureBlobLocalClient.buildClient()` results in an error (for the S3 tests it somehow [works](https://github.com/micronaut-projects/micronaut-eclipsestore/blob/1.5.x/test-suite-utils/src/main/java/io/micronaut/eclipsestore/testutils/MinioLocal.java#L49)).

### CODE REVIEW

**Comment by graemerocher** [comment]
Thanks this looks great! Could you target the 1.6.x branch since this is a new feature.

**Comment by SimonFrauenschuh** [comment]
I've switched the target branch.
Do you have any guess what could cause this blocking call?
The HttpClient is a `BlockingHttpClient` and I don't see why the `nioEventLoopGroup` would fail.
The [blocking call](https://github.com/micronaut-projects/micronaut-eclipsestore/pull/65/files#diff-e80965ef0f1a396beb49a195ceb015fe51878fa186b193a61606daf588434baeR40) that I had to move out of the [bean instantiation](https://github.com/micronaut-projects/micronaut-eclipsestore/pull/65/files#diff-adb996d87edca2a1ce25eb9324cd84a6900d557e93b79a483b9c34cb1c983cbaR21) was also leading to the same error.

According to the error logs when running `./gradlew build`
```
BlobCustomerControllerTest > testCrud(String) > [1] store FAILED
    io.micronaut.http.client.exceptions.HttpClientResponseException at DefaultHttpClient.java:2259
```
But when I just run the [BlobCustomerControllerTest](https://github.com/micronaut-projects/micronaut-eclipsestore/pull/65/files#diff-e80965ef0f1a396beb49a195ceb015fe51878fa186b193a61606daf588434bae) it seems to fail somewhere in [this](https://github.com/micronaut-projects/micronaut-eclipsestore/blob/1.5.x/eclipsestore/src/main/java/io/micronaut/eclipsestore/conf/StorageManagerFactory.java#L65) already existing function with 
```
Caused by: java.lang.IllegalStateException: block()/blockFirst()/blockLast() are blocking, which is not supported in thread default-nioEventLoopGroup-2-3
```

**Comment by graemerocher** [comment]
I think https://github.com/micronaut-projects/micronaut-eclipsestore/blob/1.5.x/test-suite/src/test/java/io/micronaut/eclipsestore/docs/CustomerController.java#L23

Needs to be annotated with `@ExecuteOn(TaskExecutors.BLOCKING)`

**Comment by SimonFrauenschuh** [comment]
Thanks!
Local, all the checks and the build process pass

**Comment by capricube** [comment]
Is it possible to add this PR to the next 4.6.0 release? 

@SimonFrauenschuh thank you for your contribution!

**Comment by graemerocher** [comment]
it is scheduled for 4.6 yes

File: eclipsestore/src/test/groovy/io/micronaut/eclipsestore/azure/BlobStorageSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,87 @@
+package io.micronaut.eclipsestore.azure
+
+import com.azure.core.credential.AzureNamedKeyCredential
```
Comment (sdelamo) [inline eclipsestore/src/test/groovy/io/micronaut/eclipsestore/azure/BlobStorageSpec.groovy]:
```suggestion
```

File: eclipsestore/src/test/groovy/io/micronaut/eclipsestore/azure/BlobStorageSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import com.azure.storage.blob.BlobServiceClient
+import com.azure.storage.blob.BlobServiceClientBuilder
+import io.micronaut.context.BeanContext
+import io.micronaut.context.annotation.Factory
+import io.micronaut.context.annotation.Property
+import io.micronaut.context.annotation.Replaces
+import io.micronaut.core.util.StringUtils
+import io.micronaut.eclipsestore.BaseStorageSpec
+import io.micronaut.eclipsestore.testutils.AzureBlobLocal
+import io.micronaut.eclipsestore.testutils.MinioLocal
```
Comment (sdelamo) [inline eclipsestore/src/test/groovy/io/micronaut/eclipsestore/azure/BlobStorageSpec.groovy]:
```suggestion
```

File: eclipsestore/src/test/groovy/io/micronaut/eclipsestore/azure/BlobStorageSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import io.micronaut.inject.qualifiers.Qualifiers
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import jakarta.inject.Inject
+import jakarta.inject.Singleton
+import org.eclipse.store.storage.embedded.types.EmbeddedStorageFoundation
+import org.eclipse.store.storage.types.StorageManager
+import org.testcontainers.DockerClientFactory
+import software.amazon.awssdk.services.s3.S3Client
```
Comment (sdelamo) [inline eclipsestore/src/test/groovy/io/micronaut/eclipsestore/azure/BlobStorageSpec.groovy]:
```suggestion
```

File: test-suite/src/test/java/io/micronaut/eclipsestore/docs/BlobCustomerControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.jupiter.api.condition.EnabledIf;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Testcontainers
+public class BlobCustomerControllerTest extends BaseCustomerControllerTest {
```
Comment (sdelamo) [inline test-suite/src/test/java/io/micronaut/eclipsestore/docs/BlobCustomerControllerTest.java]:
```suggestion
class BlobCustomerControllerTest extends BaseCustomerControllerTest {
```

**Review by sdelamo — APPROVED** [review]

**Pull request merged**